lib/ohai/mixin/azure_metadata.rb

```diff
@@ -19,30 +19,87 @@
 
 module Ohai
   module Mixin
+    #
+    # This code parses the Azure Instance Metadata API to provide details
+    # of the running instance.
+    #
+    # The code probes the instance metadata endpoint for
+    # available versions, determines the most advanced version known to
+    # work and executes the metadata retrieval using that version.
+    #
+    # If no compatible version is found, an empty hash is returned.
+    #
     module AzureMetadata
 
       AZURE_METADATA_ADDR ||= "169.254.169.254".freeze
-      AZURE_METADATA_URL ||= "/metadata/instance?api-version=2017-08-01".freeze
 
-      # fetch the meta content with a timeout and the required header
+      # it's important that the newer versions are at the end of this array so we can skip sorting it
+      AZURE_SUPPORTED_VERSIONS ||= %w{ 2017-04-02 2017-08-01 2017-12-01 2018-02-01 2018-04-02
+                                   2018-10-01 2019-02-01 2019-03-11 2019-04-30 2019-06-01
+                                   2019-06-04 2019-08-01 2019-08-15 2019-11-01 }.freeze
+
+      def best_api_version
+        @api_version ||= begin
+          logger.trace("Mixin AzureMetadata: Fetching http://#{AZURE_METADATA_ADDR}/metadata/instance to determine the latest supported metadata release")
+          response = http_get("/metadata/instance")
+          if response.code == "404"
+            logger.trace("Mixin AzureMetadata: Received HTTP 404 from metadata server while determining API version, assuming #{AZURE_SUPPORTED_VERSIONS.last}")
+            return AZURE_SUPPORTED_VERSIONS.last
+          elsif response.code != "400" # 400 is actually what we want
+            raise "Mixin AzureMetadata: Unable to determine Azure metadata version (returned #{response.code} response)"
```

Comment:
I'm really unfamiliar with the conventions of this project, but should this return a specific error type rather than the generic error?

Reply:
Most places in Ohai are just generic errors